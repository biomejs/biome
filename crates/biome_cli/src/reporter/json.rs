use crate::{DiagnosticsPayload, Execution, Reporter, ReporterVisitor, TraversalSummary};
use biome_console::fmt::Formatter;
use serde::Serialize;

#[derive(Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct JsonReporterVisitor {
    summary: TraversalSummary,
    diagnostics: Vec<biome_diagnostics::serde::Diagnostic>,
    command: String,
}

impl JsonReporterVisitor {
    pub(crate) fn new(summary: TraversalSummary) -> Self {
        Self {
            summary,
            diagnostics: vec![],
            command: String::new(),
        }
    }
}

impl biome_console::fmt::Display for JsonReporterVisitor {
    fn fmt(&self, fmt: &mut Formatter) -> std::io::Result<()> {
        let content = serde_json::to_string(&self)?;
        fmt.write_str(content.as_str())
    }
}

pub struct JsonReporter {
    pub execution: Execution,
    pub diagnostics: DiagnosticsPayload,
    pub summary: TraversalSummary,
}

impl Reporter for JsonReporter {
    fn write(self, visitor: &mut dyn ReporterVisitor) -> std::io::Result<()> {
        visitor.report_summary(&self.execution, self.summary)?;
        visitor.report_diagnostics(&self.execution, self.diagnostics)?;

        Ok(())
    }
}

impl ReporterVisitor for JsonReporterVisitor {
    fn report_summary(
        &mut self,
        execution: &Execution,
        summary: TraversalSummary,
    ) -> std::io::Result<()> {
        self.summary = summary;
        self.command = format!("{}", execution.traversal_mode());

        Ok(())
    }

    fn report_diagnostics(
        &mut self,
        _execution: &Execution,
        payload: DiagnosticsPayload,
    ) -> std::io::Result<()> {
        for diagnostic in payload.diagnostics {
            if diagnostic.severity() >= payload.diagnostic_level {
                if diagnostic.tags().is_verbose() {
                    if payload.verbose {
                        self.diagnostics
                            .push(biome_diagnostics::serde::Diagnostic::new(diagnostic))
                    }
                } else {
                    self.diagnostics
                        .push(biome_diagnostics::serde::Diagnostic::new(diagnostic))
                }
            }
        }
        Ok(())
    }
}
