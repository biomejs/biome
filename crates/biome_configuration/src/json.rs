use crate::PlainIndentStyle;
use biome_deserialize_macros::{Deserializable, Merge, Partial};
use biome_formatter::{IndentWidth, LineEnding, LineWidth};
use biome_json_formatter::context::TrailingCommas;
use bpaf::Bpaf;
use serde::{Deserialize, Serialize};

/// Options applied to JSON files
#[derive(Clone, Debug, Default, Deserialize, Eq, Partial, PartialEq, Serialize)]
#[partial(derive(Bpaf, Clone, Deserializable, Eq, Merge, PartialEq))]
#[partial(cfg_attr(feature = "schema", derive(schemars::JsonSchema)))]
#[partial(serde(default, deny_unknown_fields))]
pub struct JsonConfiguration {
    /// Parsing options
    #[partial(type, bpaf(external(partial_json_parser), optional))]
    pub parser: JsonParser,

    /// Formatting options
    #[partial(type, bpaf(external(partial_json_formatter), optional))]
    pub formatter: JsonFormatter,

    /// Linting options
    #[partial(type, bpaf(external(partial_json_linter), optional))]
    pub linter: JsonLinter,
}

/// Options that changes how the JSON parser behaves
#[derive(Clone, Debug, Default, Deserialize, Eq, Partial, PartialEq, Serialize)]
#[partial(derive(Bpaf, Clone, Deserializable, Eq, Merge, PartialEq))]
#[partial(cfg_attr(feature = "schema", derive(schemars::JsonSchema)))]
#[partial(serde(rename_all = "camelCase", default, deny_unknown_fields))]
pub struct JsonParser {
    #[partial(bpaf(hide))]
    /// Allow parsing comments in `.json` files
    pub allow_comments: bool,

    #[partial(bpaf(hide))]
    /// Allow parsing trailing commas in `.json` files
    pub allow_trailing_commas: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, Partial, PartialEq, Serialize)]
#[partial(derive(Bpaf, Clone, Deserializable, Eq, Merge, PartialEq))]
#[partial(cfg_attr(feature = "schema", derive(schemars::JsonSchema)))]
#[partial(serde(rename_all = "camelCase", default, deny_unknown_fields))]
pub struct JsonFormatter {
    /// Control the formatter for JSON (and its super languages) files.
    #[partial(bpaf(long("json-formatter-enabled"), argument("true|false"), optional))]
    pub enabled: bool,

    /// The indent style applied to JSON (and its super languages) files.
    #[partial(bpaf(long("json-formatter-indent-style"), argument("tab|space"), optional))]
    pub indent_style: Option<PlainIndentStyle>,

    /// The size of the indentation applied to JSON (and its super languages) files. Default to 2.
    #[partial(bpaf(long("json-formatter-indent-width"), argument("NUMBER"), optional))]
    pub indent_width: Option<IndentWidth>,

    /// The size of the indentation applied to JSON (and its super languages) files. Default to 2.
    #[partial(bpaf(long("json-formatter-indent-size"), argument("NUMBER"), optional))]
    #[partial(deserializable(deprecated(use_instead = "json.formatter.indentWidth")))]
    pub indent_size: Option<IndentWidth>,

    /// The type of line ending applied to JSON (and its super languages) files.
    #[partial(bpaf(long("json-formatter-line-ending"), argument("lf|crlf|cr"), optional))]
    pub line_ending: Option<LineEnding>,

    /// What's the max width of a line applied to JSON (and its super languages) files. Defaults to 80.
    #[partial(bpaf(long("json-formatter-line-width"), argument("NUMBER"), optional))]
    pub line_width: Option<LineWidth>,

    /// Print trailing commas wherever possible in multi-line comma-separated syntactic structures. Defaults to "none".
    #[partial(bpaf(long("json-formatter-trailing-commas"), argument("none|all"), optional))]
    pub trailing_commas: Option<TrailingCommas>,
}

impl PartialJsonFormatter {
    pub fn get_formatter_configuration(&self) -> JsonFormatter {
        JsonFormatter {
            enabled: self.enabled.unwrap_or_default(),
            indent_style: self.indent_style,
            indent_width: self.indent_width,
            indent_size: self.indent_size,
            line_ending: self.line_ending,
            line_width: self.line_width,
            trailing_commas: self.trailing_commas,
        }
    }
}

impl Default for JsonFormatter {
    fn default() -> Self {
        Self {
            enabled: true,
            indent_style: Default::default(),
            indent_width: Default::default(),
            indent_size: Default::default(),
            line_ending: Default::default(),
            line_width: Default::default(),
            trailing_commas: Default::default(),
        }
    }
}

/// Linter options specific to the JSON linter
#[derive(Clone, Debug, Deserialize, Eq, Partial, PartialEq, Serialize)]
#[partial(derive(Bpaf, Clone, Deserializable, Eq, Merge, PartialEq))]
#[partial(cfg_attr(feature = "schema", derive(schemars::JsonSchema)))]
#[partial(serde(rename_all = "camelCase", default, deny_unknown_fields))]
pub struct JsonLinter {
    /// Control the linter for JSON (and its super languages) files.
    #[partial(bpaf(long("json-linter-enabled"), argument("true|false"), optional))]
    pub enabled: bool,
}

impl PartialJsonFormatter {
    pub fn get_linter_configuration(&self) -> JsonLinter {
        JsonLinter {
            enabled: self.enabled.unwrap_or_default(),
        }
    }
}

impl Default for JsonLinter {
    fn default() -> Self {
        Self { enabled: true }
    }
}

impl PartialJsonLinter {
    pub fn get_linter_configuration(&self) -> JsonLinter {
        JsonLinter {
            enabled: self.enabled.unwrap_or_default(),
        }
    }
}
