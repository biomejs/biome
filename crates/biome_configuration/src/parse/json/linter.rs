use std::marker::PhantomData;

use crate::linter::{RulePlainConfiguration, RuleWithOptions};
use crate::LinterConfiguration;
use crate::RuleConfiguration;
use biome_deserialize::{
    Deserializable, DeserializableValue, DeserializationDiagnostic, DeserializationVisitor, Text,
    VisitableType,
};
use biome_rowan::TextRange;

impl Deserializable for LinterConfiguration {
    fn deserialize(
        value: &impl DeserializableValue,
        name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self> {
        value.deserialize(LinterConfigurationVisitor, name, diagnostics)
    }
}

struct LinterConfigurationVisitor;
impl DeserializationVisitor for LinterConfigurationVisitor {
    type Output = LinterConfiguration;

    const EXPECTED_TYPE: VisitableType = VisitableType::MAP;

    fn visit_map(
        self,
        members: impl Iterator<Item = Option<(impl DeserializableValue, impl DeserializableValue)>>,
        _range: biome_rowan::TextRange,
        _name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self::Output> {
        const ALLOWED_KEYS: &[&str] = &["enabled", "rules", "include", "ignore"];
        let mut result = Self::Output::default();
        for (key, value) in members.flatten() {
            let Some(key_text) = Text::deserialize(&key, "", diagnostics) else {
                continue;
            };
            match key_text.text() {
                "ignore" => {
                    result.ignore = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "include" => {
                    result.include = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "enabled" => {
                    result.enabled = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "rules" => {
                    result.rules = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                unknown_key => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
                    unknown_key,
                    key.range(),
                    ALLOWED_KEYS,
                )),
            }
        }
        Some(result)
    }
}

impl<T: Deserializable + Default> Deserializable for RuleConfiguration<T> {
    fn deserialize(
        value: &impl DeserializableValue,
        rule_name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self> {
        value.deserialize(
            RuleConfigurationVisitor(PhantomData),
            rule_name,
            diagnostics,
        )
    }
}

struct RuleConfigurationVisitor<T>(PhantomData<T>);
impl<T: Deserializable + Default> DeserializationVisitor for RuleConfigurationVisitor<T> {
    type Output = RuleConfiguration<T>;

    const EXPECTED_TYPE: VisitableType = VisitableType::STR.union(VisitableType::MAP);

    fn visit_str(
        self,
        value: Text,
        range: TextRange,
        _rule_name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self::Output> {
        RulePlainConfiguration::deserialize_from_str(value, range, diagnostics)
            .map(RuleConfiguration::Plain)
    }

    fn visit_map(
        self,
        members: impl Iterator<Item = Option<(impl DeserializableValue, impl DeserializableValue)>>,
        _range: biome_rowan::TextRange,
        rule_name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self::Output> {
        const ALLOWED_KEYS: &[&str] = &["level", "options"];
        let mut result = RuleWithOptions::default();
        for (key, value) in members.flatten() {
            let Some(key_text) = Text::deserialize(&key, "", diagnostics) else {
                continue;
            };
            match key_text.text() {
                "level" => {
                    result.level = Deserializable::deserialize(&value, &key_text, diagnostics)?;
                }
                "options" => {
                    if let Some(options) =
                        Deserializable::deserialize(&value, rule_name, diagnostics)
                    {
                        result.options = options;
                    }
                }
                unknown_key => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
                    unknown_key,
                    key.range(),
                    ALLOWED_KEYS,
                )),
            }
        }
        Some(RuleConfiguration::WithOptions(result))
    }
}

impl RulePlainConfiguration {
    fn deserialize_from_str(
        value: Text,
        range: TextRange,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self> {
        const ALLOWED_VARIANTS: &[&str] = &["error", "warn", "off"];
        if let Ok(value) = value.text().parse::<Self>() {
            Some(value)
        } else {
            diagnostics.push(DeserializationDiagnostic::new_unknown_value(
                value.text(),
                range,
                ALLOWED_VARIANTS,
            ));
            None
        }
    }
}

impl Deserializable for RulePlainConfiguration {
    fn deserialize(
        value: &impl DeserializableValue,
        name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self> {
        let value_text = Text::deserialize(value, name, diagnostics)?;
        Self::deserialize_from_str(value_text, value.range(), diagnostics)
    }
}
