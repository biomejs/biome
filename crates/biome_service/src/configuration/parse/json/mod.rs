//! This module is responsible to parse the configuration from a JSON format
//!

mod configuration;
mod files;
mod formatter;
mod javascript;
mod json_impl;
mod linter;
mod organize_imports;
mod overrides;
mod rules;
mod vcs;
