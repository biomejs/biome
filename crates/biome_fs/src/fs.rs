use crate::{BiomePath, PathInterner};
use biome_diagnostics::{console, Advices, Diagnostic, LogCategory, Visit};
use biome_diagnostics::{Error, Severity};
pub use memory::{ErrorEntry, MemoryFileSystem};
pub use os::OsFileSystem;
use oxc_resolver::{Resolution, ResolveError};
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use std::panic::RefUnwindSafe;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::{fmt, io};
use tracing::{error, info};

mod memory;
mod os;

pub const ROME_JSON: &str = "rome.json";

pub struct ConfigName;

impl ConfigName {
    const BIOME_JSON: [&'static str; 2] = ["biome.json", "biome.jsonc"];

    pub const fn biome_json() -> &'static str {
        Self::BIOME_JSON[0]
    }

    pub const fn biome_jsonc() -> &'static str {
        Self::BIOME_JSON[1]
    }

    pub const fn file_names() -> [&'static str; 2] {
        Self::BIOME_JSON
    }
}

type AutoSearchResultAlias = Result<Option<AutoSearchResult>, FileSystemDiagnostic>;

pub trait FileSystem: Send + Sync + RefUnwindSafe {
    /// It opens a file with the given set of options
    fn open_with_options(&self, path: &Path, options: OpenOptions) -> io::Result<Box<dyn File>>;

    /// Initiate a traversal of the filesystem
    ///
    /// This method creates a new "traversal scope" that can be used to
    /// efficiently batch many filesystem read operations
    fn traversal<'scope>(&'scope self, func: BoxedTraversal<'_, 'scope>);

    // TODO: remove once we remove `rome.json` support [2.0]
    /// Returns the temporary configuration files that are supported
    fn deprecated_config_name(&self) -> &str {
        ROME_JSON
    }

    /// Return the path to the working directory
    fn working_directory(&self) -> Option<PathBuf>;

    /// Checks if the given path exists in the file system
    fn path_exists(&self, path: &Path) -> bool;

    /// Checks if the given path is a regular file
    fn path_is_file(&self, path: &Path) -> bool;

    /// Checks if the given path is a directory
    fn path_is_dir(&self, path: &Path) -> bool;

    /// Checks if the given path is a symlink
    fn path_is_symlink(&self, path: &Path) -> bool;

    /// Method that takes a path to a folder `file_path`, and a `file_name`. It attempts to find
    /// and read the file from that folder and if not found, it reads the parent directories recursively
    /// until:
    /// - the file is found, then it reads and return its contents
    /// - the file is not found
    ///
    /// If `should_error_if_file_not_found` it `true`, it returns an error.
    ///
    /// ## Errors
    ///
    /// - The file can't be read
    ///
    fn auto_search(
        &self,
        mut search_dir: PathBuf,
        filenames: &[&str],
        should_error_if_file_not_found: bool,
    ) -> AutoSearchResultAlias {
        let mut is_searching_in_parent_dir = false;
        loop {
            let mut errors: Vec<(FileSystemDiagnostic, String)> = vec![];

            for filename in filenames {
                let file_path = search_dir.join(filename);
                let open_options = OpenOptions::default().read(true);
                match self.open_with_options(&file_path, open_options) {
                    Ok(mut file) => {
                        let mut buffer = String::new();
                        match file.read_to_string(&mut buffer) {
                            Ok(_) => {
                                if is_searching_in_parent_dir {
                                    info!(
                                        "Biome auto discovered the file at following path that wasn't in the working directory: {}",
                                        search_dir.display()
                                    );
                                }
                                return Ok(Some(AutoSearchResult {
                                    content: buffer,
                                    file_path,
                                    directory_path: search_dir,
                                }));
                            }
                            Err(err) => {
                                if !is_searching_in_parent_dir && should_error_if_file_not_found {
                                    let error_message = format!(
                                        "Biome couldn't read the file {:?}, reason:\n {}",
                                        file_path, err
                                    );
                                    errors.push((
                                        FileSystemDiagnostic {
                                            path: file_path.display().to_string(),
                                            severity: Severity::Error,
                                            error_kind: ErrorKind::CantReadFile(
                                                file_path.display().to_string(),
                                            ),
                                        },
                                        error_message,
                                    ));
                                }
                                continue;
                            }
                        }
                    }
                    Err(err) => {
                        if !is_searching_in_parent_dir && should_error_if_file_not_found {
                            let error_message = format!(
                                "Biome couldn't find the file {:?}, reason:\n {}",
                                file_path, err
                            );
                            errors.push((
                                FileSystemDiagnostic {
                                    path: file_path.display().to_string(),
                                    severity: Severity::Error,
                                    error_kind: ErrorKind::CantReadFile(
                                        file_path.display().to_string(),
                                    ),
                                },
                                error_message,
                            ));
                        }
                        continue;
                    }
                }
            }

            if !is_searching_in_parent_dir && should_error_if_file_not_found {
                if let Some(diagnostic) = errors
                    .into_iter()
                    .map(|(diagnostic, message)| {
                        error!(message);
                        diagnostic
                    })
                    .last()
                {
                    // we can only return one Err, so we return the last diagnostic.
                    return Err(diagnostic);
                }
            }

            if let Some(parent_search_dir) = search_dir.parent() {
                search_dir = PathBuf::from(parent_search_dir);
                is_searching_in_parent_dir = true;
            } else {
                break;
            }
        }

        Ok(None)
    }

    fn get_changed_files(&self, base: &str) -> io::Result<Vec<String>>;

    fn get_staged_files(&self) -> io::Result<Vec<String>>;

    fn resolve_configuration(
        &self,
        specifier: &str,
        path: Option<&Path>,
    ) -> Result<Resolution, ResolveError>;
}

/// Result of the auto search
#[derive(Debug)]
pub struct AutoSearchResult {
    /// The content of the file
    pub content: String,
    /// The path of the directory where the file was found
    pub directory_path: PathBuf,
    /// The path of the file found
    pub file_path: PathBuf,
}

pub trait File {
    /// Read the content of the file into `buffer`
    fn read_to_string(&mut self, buffer: &mut String) -> io::Result<()>;

    /// Overwrite the content of the file with the provided bytes
    ///
    /// This will write to the associated memory buffer, as well as flush the
    /// new content to the disk if this is a physical file
    fn set_content(&mut self, content: &[u8]) -> io::Result<()>;

    /// Returns the version of the current file
    fn file_version(&self) -> i32;
}

/// This struct is a "mirror" of [std::fs::FileOptions].
/// Refer to their documentation for more details
#[derive(Default, Debug)]
pub struct OpenOptions {
    read: bool,
    write: bool,
    truncate: bool,
    create: bool,
    create_new: bool,
}

impl OpenOptions {
    pub fn read(mut self, read: bool) -> Self {
        self.read = read;
        self
    }
    pub fn write(mut self, write: bool) -> Self {
        self.write = write;
        self
    }
    pub fn truncate(mut self, truncate: bool) -> Self {
        self.truncate = truncate;
        self
    }
    pub fn create(mut self, create: bool) -> Self {
        self.create = create;
        self
    }
    pub fn create_new(mut self, create_new: bool) -> Self {
        self.create_new = create_new;
        self
    }

    pub fn into_fs_options(self, options: &mut std::fs::OpenOptions) -> &mut std::fs::OpenOptions {
        options
            .read(self.read)
            .write(self.write)
            .truncate(self.truncate)
            .create(self.create)
            .create_new(self.create_new)
    }
}

/// Trait that contains additional methods to work with [FileSystem]
pub trait FileSystemExt: FileSystem {
    /// Open a file with the `read` option
    ///
    /// Equivalent to [std::fs::File::open]
    fn open(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(path, OpenOptions::default().read(true))
    }

    /// Open a file with the `write` and `create` options
    ///
    /// Equivalent to [std::fs::File::create]
    fn create(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(
            path,
            OpenOptions::default()
                .write(true)
                .create(true)
                .truncate(true),
        )
    }

    /// Opens a file with the `read`, `write` and `create_new` options
    ///
    /// Equivalent to [std::fs::File::create_new]
    fn create_new(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(
            path,
            OpenOptions::default()
                .read(true)
                .write(true)
                .create_new(true),
        )
    }
}

impl<T: ?Sized> FileSystemExt for T where T: FileSystem {}

type BoxedTraversal<'fs, 'scope> = Box<dyn FnOnce(&dyn TraversalScope<'scope>) + Send + 'fs>;

pub trait TraversalScope<'scope> {
    /// Spawn a new filesystem read task
    ///
    /// If the provided path exists and is a file, then the [`handle_file`](TraversalContext::handle_file)
    /// method of the provided [TraversalContext] will be called. If it's a
    /// directory, it will be recursively traversed and all the files the
    /// [`can_handle`](TraversalContext::can_handle) method of the context
    /// returns true for will be handled as well
    fn spawn(&self, context: &'scope dyn TraversalContext, path: PathBuf);
}

pub trait TraversalContext: Sync {
    /// Provides the traversal scope with an instance of [PathInterner], used
    /// to emit diagnostics for IO errors that may happen in the traversal process
    fn interner(&self) -> &PathInterner;

    /// Called by the traversal process to emit an error diagnostic associated
    /// with a particular file ID when an IO error happens
    fn push_diagnostic(&self, error: Error);

    /// Checks if the traversal context can handle a particular path, used as
    /// an optimization to bail out of scheduling a file handler if it wouldn't
    /// be able to process the file anyway
    fn can_handle(&self, path: &BiomePath) -> bool;

    /// This method will be called by the traversal for each file it finds
    /// where [TraversalContext::can_handle] returned true
    fn handle_file(&self, path: &Path);
}

impl<T> FileSystem for Arc<T>
where
    T: FileSystem + Send,
{
    fn open_with_options(&self, path: &Path, options: OpenOptions) -> io::Result<Box<dyn File>> {
        T::open_with_options(self, path, options)
    }

    fn traversal<'scope>(&'scope self, func: BoxedTraversal<'_, 'scope>) {
        T::traversal(self, func)
    }

    fn working_directory(&self) -> Option<PathBuf> {
        T::working_directory(self)
    }

    fn path_exists(&self, path: &Path) -> bool {
        T::path_exists(self, path)
    }

    fn path_is_file(&self, path: &Path) -> bool {
        T::path_is_file(self, path)
    }

    fn path_is_dir(&self, path: &Path) -> bool {
        T::path_is_dir(self, path)
    }

    fn path_is_symlink(&self, path: &Path) -> bool {
        T::path_is_symlink(self, path)
    }

    fn get_changed_files(&self, base: &str) -> io::Result<Vec<String>> {
        T::get_changed_files(self, base)
    }

    fn get_staged_files(&self) -> io::Result<Vec<String>> {
        T::get_staged_files(self)
    }

    fn resolve_configuration(
        &self,
        specifier: &str,
        path: Option<&Path>,
    ) -> Result<Resolution, ResolveError> {
        T::resolve_configuration(self, specifier, path)
    }
}

#[derive(Debug, Diagnostic, Deserialize, Serialize)]
#[diagnostic(category = "internalError/fs")]
pub struct FileSystemDiagnostic {
    #[severity]
    pub severity: Severity,
    #[location(resource)]
    pub path: String,
    #[message]
    #[description]
    #[advice]
    pub error_kind: ErrorKind,
}

impl Display for FileSystemDiagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        Diagnostic::description(self, f)
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum ErrorKind {
    /// File not found
    CantReadFile(String),
    /// Unknown file type
    UnknownFileType,
    /// Dereferenced (broken) symbolic link
    DereferencedSymlink(String),
    /// Too deeply nested symbolic link expansion
    DeeplyNestedSymlinkExpansion(String),
}

impl console::fmt::Display for ErrorKind {
    fn fmt(&self, fmt: &mut console::fmt::Formatter) -> io::Result<()> {
        match self {
            ErrorKind::CantReadFile(_) => fmt.write_str("Biome couldn't read the file"),
            ErrorKind::UnknownFileType => fmt.write_str("Unknown file type"),
            ErrorKind::DereferencedSymlink(_) => fmt.write_str("Dereferenced symlink"),
            ErrorKind::DeeplyNestedSymlinkExpansion(_) => {
                fmt.write_str("Deeply nested symlink expansion")
            }
        }
    }
}

impl std::fmt::Display for ErrorKind {
    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ErrorKind::CantReadFile(_) => fmt.write_str("Biome couldn't read the file"),
            ErrorKind::UnknownFileType => write!(fmt, "Unknown file type"),
            ErrorKind::DereferencedSymlink(_) => write!(fmt, "Dereferenced symlink"),
            ErrorKind::DeeplyNestedSymlinkExpansion(_) => {
                write!(fmt, "Deeply nested symlink expansion")
            }
        }
    }
}

impl Advices for ErrorKind {
    fn record(&self, visitor: &mut dyn Visit) -> io::Result<()> {
        match self {
			ErrorKind::CantReadFile(path) => visitor.record_log(
		LogCategory::Error,
			&format!("Biome couldn't read the following file, maybe for permissions reasons or it doesn't exists: {}", path)
			),

            ErrorKind::UnknownFileType => visitor.record_log(
                LogCategory::Info,
                &"Biome encountered a file system entry that's neither a file, directory or symbolic link",
            ),
            ErrorKind::DereferencedSymlink(path) => visitor.record_log(
                LogCategory::Info,
                &format!("Biome encountered a file system entry that is a broken symbolic link: {}", path),
            ),
            ErrorKind::DeeplyNestedSymlinkExpansion(path) => visitor.record_log(
                LogCategory::Error,
                &format!("Biome encountered a file system entry with too many nested symbolic links, possibly forming an infinite cycle: {}", path),
            ),
        }
    }
}
