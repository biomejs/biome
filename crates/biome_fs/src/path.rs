//! This module is responsible to manage paths inside Biome.
//! It is a small wrapper around [path::PathBuf] but it is also able to
//! give additional information around the file that holds:
//! - the [FileHandlers] for the specific file
//! - shortcuts to open/write to the file
use std::ffi::OsStr;
use std::fs::{self, read_to_string};
use std::{fs::File, io, io::Write, ops::Deref, path::PathBuf};

#[derive(Debug, Clone, Eq, Hash, PartialEq, Default)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
pub struct BiomePath {
    path: PathBuf,
}

impl Deref for BiomePath {
    type Target = PathBuf;

    fn deref(&self) -> &Self::Target {
        &self.path
    }
}

impl BiomePath {
    pub fn new(path_to_file: impl Into<PathBuf>) -> Self {
        Self {
            path: path_to_file.into(),
        }
    }

    /// Accepts a file opened in read mode and saves into it
    pub fn save(&mut self, content: &str) -> Result<(), std::io::Error> {
        let mut file_to_write = File::create(&self.path).unwrap();
        // TODO: handle error with diagnostic
        file_to_write.write_all(content.as_bytes())
    }

    /// Returns the contents of a file, if it exists
    pub fn get_buffer_from_file(&mut self) -> String {
        // we assume we have permissions
        fs::read_to_string(&self.path).expect("cannot read the file to format")
    }

    /// Small wrapper for [read_to_string]
    pub fn read_to_string(&self) -> io::Result<String> {
        let path = self.path.as_path();
        read_to_string(path)
    }

    pub fn extension_as_str(&self) -> Option<&str> {
        self.extension().and_then(OsStr::to_str)
    }
}
