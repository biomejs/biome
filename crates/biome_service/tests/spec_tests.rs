use biome_deserialize::json::deserialize_from_json_str;
use biome_diagnostics::{print_diagnostic_to_string, DiagnosticExt};
use biome_json_parser::JsonParserOptions;
use biome_service::Configuration;
use std::ffi::OsStr;
use std::fs::read_to_string;
use std::path::Path;

tests_macros::gen_tests! {"tests/invalid/*.{json}", crate::run_invalid_configurations, "module"}
tests_macros::gen_tests! {"tests/valid/*.{json}", crate::run_valid_configurations, "module"}

fn run_invalid_configurations(input: &'static str, _: &str, _: &str, _: &str) {
    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let extension = input_file.extension().and_then(OsStr::to_str).unwrap();
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let result = match extension {
        "json" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default(),
        ),
        "jsonc" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default().with_allow_comments(),
        ),
        _ => {
            panic!("Extension not supported");
        }
    };

    assert!(
        result.has_errors(),
        "This test should have diagnostics, but it doesn't have any"
    );

    let result = result
        .into_diagnostics()
        .into_iter()
        .map(|diagnostic| {
            print_diagnostic_to_string(
                &diagnostic
                    .with_file_path(file_name)
                    .with_file_source_code(input_code.as_str()),
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n");

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, result, file_name);
    });
}

fn run_valid_configurations(input: &'static str, _: &str, _: &str, _: &str) {
    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let extension = input_file.extension().and_then(OsStr::to_str).unwrap();
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let result = match extension {
        "json" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default(),
        ),
        "jsonc" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default().with_allow_comments(),
        ),
        _ => {
            panic!("Extension not supported");
        }
    };

    let has_errors = result.has_errors();

    if has_errors {
        let diagnostics = result
            .into_diagnostics()
            .into_iter()
            .map(|diagnostic| {
                print_diagnostic_to_string(
                    &diagnostic
                        .with_file_path(file_name)
                        .with_file_source_code(input_code.as_str()),
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n");
        panic!(
            "This test should not have diagnostics, but some have been emitted.\n {}",
            diagnostics
        );
    } else {
        // test passed, no errors
    }
}

#[ignore]
#[test]
fn quick_test() {
    let source = r#"{
        "javascript": {
            "formatter": {
                "overrides": [
                {}]
            }
        }
    }"#;
    let result = deserialize_from_json_str::<Configuration>(source, JsonParserOptions::default());

    dbg!(result.diagnostics());
    assert!(!result.has_errors());
}
