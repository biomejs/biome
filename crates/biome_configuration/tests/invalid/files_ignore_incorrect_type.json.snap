---
source: crates/biome_service/tests/spec_tests.rs
expression: files_ignore_incorrect_type.json
---
files_ignore_incorrect_type.json:3:13 deserialize ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × ignore has an incorrect type, expected an array, but received a string.
  
    1 │ {
    2 │ 	"files": {
  > 3 │ 		"ignore": "something"
      │ 		          ^^^^^^^^^^^
    4 │ 	}
    5 │ }
  


